<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_20) on Wed Sep 22 13:35:10 GFT 2010 -->
<TITLE>
DNASeq
</TITLE>

<META NAME="date" CONTENT="2010-09-22">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="DNASeq";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/DNASeq.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../sequences/Apilable.html" title="interface in sequences"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../sequences/DNASeqTest.html" title="class in sequences"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../index.html?sequences/DNASeq.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="DNASeq.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
sequences</FONT>
<BR>
Class DNASeq</H2>
<PRE>
java.lang.Object
  <IMG SRC="../resources/inherit.gif" ALT="extended by "><A HREF="../sequences/Sequence.html" title="class in sequences">sequences.Sequence</A>
      <IMG SRC="../resources/inherit.gif" ALT="extended by "><B>sequences.DNASeq</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../sequences/Apilable.html" title="interface in sequences">Apilable</A></DD>
</DL>
<DL>
<DT><B>Direct Known Subclasses:</B> <DD><A HREF="../sequences/Primer.html" title="class in sequences">Primer</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>DNASeq</B><DT>extends <A HREF="../sequences/Sequence.html" title="class in sequences">Sequence</A><DT>implements <A HREF="../sequences/Apilable.html" title="interface in sequences">Apilable</A></DL>
</PRE>

<P>
Class to represent a sequence of DNA. (or RNA)
<P>

<P>
<DL>
<DT><B>Version:</B></DT>
  <DD>1.1.7</DD>
<DT><B>Author:</B></DT>
  <DD>Javier Iserte <jiserte@unq.edu.ar></DD>
</DL>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../sequences/DNASeq.html#DNASeq(java.lang.String, java.lang.String)">DNASeq</A></B>(java.lang.String&nbsp;sequence,
       java.lang.String&nbsp;description)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../sequences/DNASeq.html#complementary(java.lang.String)">complementary</A></B>(java.lang.String&nbsp;sequence)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the complementary strand of a DNA sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../sequences/Primer.html" title="class in sequences">Primer</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../sequences/DNASeq.html#designPrimer(int, int, boolean)">designPrimer</A></B>(int&nbsp;from,
             int&nbsp;to,
             boolean&nbsp;directStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This method is used to design a primer from a DNA sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../sequences/DNASeq.html#getComplementary()">getComplementary</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../sequences/DNASeq.html#pileUpWith(sequences.Apilable, degeneration.GeneticCode)">pileUpWith</A></B>(<A HREF="../sequences/Apilable.html" title="interface in sequences">Apilable</A>&nbsp;anotherApilable,
           <A HREF="../degeneration/GeneticCode.html" title="class in degeneration">GeneticCode</A>&nbsp;myGC)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method is used to pile up a DNA sequence with another sequence into a DNA sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../sequences/DNASeq.html#pileUpWithDNAseq(sequences.DNASeq, degeneration.GeneticCode)">pileUpWithDNAseq</A></B>(<A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A>&nbsp;anotherSeq,
                 <A HREF="../degeneration/GeneticCode.html" title="class in degeneration">GeneticCode</A>&nbsp;myGC)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method is used to pile up two DNA sequences into a DNA sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../sequences/DNASeq.html#pileUpWithProtseq(sequences.ProtSeq, degeneration.GeneticCode)">pileUpWithProtseq</A></B>(<A HREF="../sequences/ProtSeq.html" title="class in sequences">ProtSeq</A>&nbsp;anotherSeq,
                  <A HREF="../degeneration/GeneticCode.html" title="class in degeneration">GeneticCode</A>&nbsp;myGC)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method is used to pile up a DNA sequence to a protein sequence into a DNA sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../sequences/ProtSeq.html" title="class in sequences">ProtSeq</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../sequences/DNASeq.html#traducir(degeneration.GeneticCode)">traducir</A></B>(<A HREF="../degeneration/GeneticCode.html" title="class in degeneration">GeneticCode</A>&nbsp;myGc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This method translate a DNA sequence into a amino acid sequence, according to a genetic Code.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_sequences.Sequence"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class sequences.<A HREF="../sequences/Sequence.html" title="class in sequences">Sequence</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../sequences/Sequence.html#getDescription()">getDescription</A>, <A HREF="../sequences/Sequence.html#getLength()">getLength</A>, <A HREF="../sequences/Sequence.html#getSequence()">getSequence</A>, <A HREF="../sequences/Sequence.html#setDescription(java.lang.String)">setDescription</A>, <A HREF="../sequences/Sequence.html#setSequence(java.lang.String)">setSequence</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="DNASeq(java.lang.String, java.lang.String)"><!-- --></A><H3>
DNASeq</H3>
<PRE>
public <B>DNASeq</B>(java.lang.String&nbsp;sequence,
              java.lang.String&nbsp;description)</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getComplementary()"><!-- --></A><H3>
getComplementary</H3>
<PRE>
public <A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A> <B>getComplementary</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="traducir(degeneration.GeneticCode)"><!-- --></A><H3>
traducir</H3>
<PRE>
public <A HREF="../sequences/ProtSeq.html" title="class in sequences">ProtSeq</A> <B>traducir</B>(<A HREF="../degeneration/GeneticCode.html" title="class in degeneration">GeneticCode</A>&nbsp;myGc)</PRE>
<DL>
<DD>This method translate a DNA sequence into a amino acid sequence, according to a genetic Code.
 
 Precondition: The sequence length is multiple of three.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>myGc</CODE> - a genetic code.
<DT><B>Returns:</B><DD>a Protein sequence</DL>
</DD>
</DL>
<HR>

<A NAME="pileUpWith(sequences.Apilable, degeneration.GeneticCode)"><!-- --></A><H3>
pileUpWith</H3>
<PRE>
public <A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A> <B>pileUpWith</B>(<A HREF="../sequences/Apilable.html" title="interface in sequences">Apilable</A>&nbsp;anotherApilable,
                         <A HREF="../degeneration/GeneticCode.html" title="class in degeneration">GeneticCode</A>&nbsp;myGC)</PRE>
<DL>
<DD>this method is used to pile up a DNA sequence with another sequence into a DNA sequence.
 Is part of a 'double dispatching' pattern.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../sequences/Apilable.html#pileUpWith(sequences.Apilable, degeneration.GeneticCode)">pileUpWith</A></CODE> in interface <CODE><A HREF="../sequences/Apilable.html" title="interface in sequences">Apilable</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="pileUpWithDNAseq(sequences.DNASeq, degeneration.GeneticCode)"><!-- --></A><H3>
pileUpWithDNAseq</H3>
<PRE>
public <A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A> <B>pileUpWithDNAseq</B>(<A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A>&nbsp;anotherSeq,
                               <A HREF="../degeneration/GeneticCode.html" title="class in degeneration">GeneticCode</A>&nbsp;myGC)</PRE>
<DL>
<DD>this method is used to pile up two DNA sequences into a DNA sequence.
 Is part of a 'double dispatching' pattern.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../sequences/Apilable.html#pileUpWithDNAseq(sequences.DNASeq, degeneration.GeneticCode)">pileUpWithDNAseq</A></CODE> in interface <CODE><A HREF="../sequences/Apilable.html" title="interface in sequences">Apilable</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="pileUpWithProtseq(sequences.ProtSeq, degeneration.GeneticCode)"><!-- --></A><H3>
pileUpWithProtseq</H3>
<PRE>
public <A HREF="../sequences/DNASeq.html" title="class in sequences">DNASeq</A> <B>pileUpWithProtseq</B>(<A HREF="../sequences/ProtSeq.html" title="class in sequences">ProtSeq</A>&nbsp;anotherSeq,
                                <A HREF="../degeneration/GeneticCode.html" title="class in degeneration">GeneticCode</A>&nbsp;myGC)</PRE>
<DL>
<DD>this method is used to pile up a DNA sequence to a protein sequence into a DNA sequence.
 Is part of a 'double dispatching' pattern.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../sequences/Apilable.html#pileUpWithProtseq(sequences.ProtSeq, degeneration.GeneticCode)">pileUpWithProtseq</A></CODE> in interface <CODE><A HREF="../sequences/Apilable.html" title="interface in sequences">Apilable</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="designPrimer(int, int, boolean)"><!-- --></A><H3>
designPrimer</H3>
<PRE>
public <A HREF="../sequences/Primer.html" title="class in sequences">Primer</A> <B>designPrimer</B>(int&nbsp;from,
                           int&nbsp;to,
                           boolean&nbsp;directStrand)</PRE>
<DL>
<DD>This method is used to design a primer from a DNA sequence. 
 The first position of the sequence is One and the last is the same that the length of the sequence.
 If directStrans is false, it is assumed that this sequence is the complementary of an original one.
 In this case the value 'from' and 'to' are modified to fit the original sequence.
 If you want to design primers in the complementary strand, you only need to create it just once.
 
 <blockquote>
 Example: <br>
 My original sequence 'mySeq' is:<br>
 <p><code>
 |---------- ----------21-------30 ---------- ---------- ----------<br>  
 "AAAAAAAAAA AAAAAAAAAA GCAGCAGCAG AAAAAAAAAA AAAAAAAAAA AAAAAAAAAA"<br>
 "TTTTTTTTTT TTTTTTTTTT CGTCGTCGTC TTTTTTTTTT TTTTTTTTTT TTTTTTTTTT"<br>
  </code></p>
 I want to design a primer from position 21 to 30 in the direct strand (GCAGCAGCAG).:
 <p><code>
 Primer myPrimer = mySeq.designPrimer(21,30,true)<br>
 </code></p>
  
 If I want to design a primer from position 21 to 30 in the complementary strand of 'mySeq' (CTGCTGCTGC):
 <p><code>
 myCompSeq = mySeq.getComplementary();<br>
 Primer myPrimer = myCompSeq.designPrimer(21,30,false);
  </code></p>
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>from</CODE> - an index that indicates the first position of the primer<DD><CODE>to</CODE> - an index that indicates the last position of the primer
<DT><B>Returns:</B><DD>Primer the resulting primer.</DL>
</DD>
</DL>
<HR>

<A NAME="complementary(java.lang.String)"><!-- --></A><H3>
complementary</H3>
<PRE>
public java.lang.String <B>complementary</B>(java.lang.String&nbsp;sequence)</PRE>
<DL>
<DD>Gets the complementary strand of a DNA sequence.
 This method supports the IUPAC ambiguity code for bases.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>sequence</CODE> - the original DNA sequence.
<DT><B>Returns:</B><DD>another DNA Sequence that is the complementary one.</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/DNASeq.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../sequences/Apilable.html" title="interface in sequences"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../sequences/DNASeqTest.html" title="class in sequences"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../index.html?sequences/DNASeq.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="DNASeq.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
